boot/assets.go

```diff
@@ -0,0 +1,59 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/gadget"
+)
+
+func InitialSealObserver(model *asserts.Model) gadget.ContentWriteObserver {
+	if model.Grade() == asserts.ModelGradeUnset {
+		// no need to observe updates when assets are not managed
+		return nil
+	}
+
+	return &initialSealObserver{
+		model: model,
+	}
+}
+
+type initialSealObserver struct {
+	model *asserts.Model
+}
+
+// Implements gadget.FileWriteObserver.
+func (o *initialSealObserver) Observe(op gadget.ObserveAction, affectedStruct *gadget.LaidOutStructure, root, realSource, relativeTarget string) (gadget.ObserveResult, error) {
+	// TODO:UC20:
+	// steps on write action:
+	// - copy new asset to assets cache
+	// - update modeeenv
+	// steps on rollback action:
+	// - drop file from cache if no longer referenced
+	// - update modeenv
+	return gadget.ObserveResultNoted, nil
+}
+
+// Implements gadget.FileWriteObserver.
+func (o *initialSealObserver) Apply() error {
+	// TODO:UC20: steps:
+	// - reseal
```

Comment:
That's a fair point. I think that we should do the analogous thing to we discussed for the gadget update case with @pedronis, that is, call Apply after the writers are done. I left a note for it right here https://github.com/snapcore/snapd/pull/9089/commits/66fe706b62d34edfa89ac3e96da0672ddb6fa54f#diff-7c4a0d9bd3c13f6d1c75144f29cafdc6R127 We'd have to do the call in install.Run(), since we know at this point when all content has been populated.